testnet/stacks-node/src/nakamoto_node/miner.rs

```diff
@@ -241,6 +270,90 @@ impl BlockMinerThread {
+    ) -> Option<NakamotoBlock> {
+        let stackerdb_contracts = stackerdbs
+            .get_stackerdb_contract_ids()
+            .expect("FATAL: could not get the stacker DB contract ids");
+        // TODO: get this directly instead of this jankiness when .signers is a boot contract
+        let signers_contract_id = boot_code_id(SIGNERS_NAME, self.config.is_mainnet());
+        if !stackerdb_contracts.contains(&signers_contract_id) {
+            debug!("No signers contract found, cannot wait for signers");
+            return None;
+        };
+        // Get the block slot for every signer
+        let slot_ids = stackerdbs
+            .get_signers(&signers_contract_id)
+            .expect("FATAL: could not get signers from stacker DB")
+            .iter()
+            .enumerate()
+            .map(|(id, _)| id as u32 * SIGNER_SLOTS_PER_USER + BLOCK_SLOT_ID)
+            .collect::<Vec<u32>>();
+        // If more than a threshold percentage of the signers reject the block, we should not wait any further
+        let rejection_threshold =
+            slot_ids.len() / 100 * self.config.miner.signer_rejection_threshold;
+        let mut num_rejections = 0;
+        let now = Instant::now();
+        while now.elapsed() < self.config.miner.wait_on_signers {
+            // Get the block responses from the signers for the block we just proposed
+            let signer_messages: Vec<SignerMessage> = stackerdbs
+                .get_latest_chunks(&signers_contract_id, &slot_ids)
+                .expect("FATAL: could not get latest chunks from stacker DB")
+                .into_iter()
+                .filter_map(|chunk| {
+                    chunk.and_then(|chunk| bincode::deserialize::<SignerMessage>(&chunk).ok())
+                })
+                .collect();
+            for signer_message in signer_messages {
+                match signer_message {
+                    SignerMessage::BlockResponse(BlockResponse::Accepted(block)) => {
+                        if block.header.signer_signature_hash().ok()? == *signer_signature_hash
+                            && block
+                                .header
+                                .signer_signature
+                                .verify(aggregate_public_key, &signer_signature_hash.0)
+                        {
+                            // We verified that the returned block contains a valid signature across the signer signature hash of our original block request
+                            // Let's just immediately attempt to append the block without waiting for more signer messages
+                            return Some(block);
+                        }
+                    }
+                    SignerMessage::BlockResponse(BlockResponse::Rejected(block_rejection)) => {
+                        // First check that this block rejection is for the block we proposed
+                        if block_rejection.block.header.signer_signature_hash().ok()?
```

Comment:
It will continue to wait for verification and not prematurely mine a new block (not until enough rejections appear that are valid indicating that the block is not gonna get accepted.) They will then mine a new block.